zookeeper-common/src/main/java/org/apache/zookeeper/ClientCnxnSocketNetty.java

```diff
@@ -184,7 +213,9 @@ void cleanup() {
 
     @Override
     void close() {
-        channelFactory.releaseExternalResources();
+        if (!eventLoopGroup.isShuttingDown()) {
```

Comment:
Is this really needed?

Reply:
I'm not sure if calling `shutdownGracefully` more than once is allowed, which is why I added the check. It might not be necessary.